Only query the requested path in the file API

mvanderkamp opened the pull request:
This is to follow up on discussion #1866 

Currently this is still just the first step as discussed, but I figured a PR is a better place to look at and discuss actual code than a discussion post.

I basically just extracted the relevant lines from the body of `TreeInfo`. I did also leave in a call to `TreeInfo` for the `if empty(path)` case though, since it seemed to me like the commit could still be either a regular hash or a stage number in that case, which get the ftime in different ways.

tpope commented on autoload/fugitive.vim:
```diff
@@ -2044,8 +2044,24 @@ function! s:PathInfo(url) abort
+    let [_, ftime] = s:TreeInfo(dir, commit)
+    let entry = [ftime, '040000', 'tree', '', -1]
+  elseif commit =~# '^:\=[0-3]$'
+    let [lines, exec_error] = s:LinesError([dir, 'ls-files', '--stage', '--', path])
```
```vim
let result = fugitive#Execute(['--literal-pathspecs', 'ls-files', '--stage', '--', path])
let line = result.stdout[0]
if result.exit_status || empty(line)
```

The `--literal-pathspecs` is a necessary change. The rest is just modernization.

mvanderkamp commented on autoload/fugitive.vim:
```diff
@@ -2044,8 +2044,24 @@ function! s:PathInfo(url) abort
+      return [-1, '000000', '', '', -1]
+    endif
+    let newftime = getftime(fugitive#Find('.git/index', dir))
+    let [info, filename] = split(lines[0], "\t")
```
Ach, sorry! Somewhere in there while tinkering with the output of `ls-files` and `ls-tree` I somehow convinced myself that the directory would also be listed by `ls-files` as well as its contents. I'll take another crack at it.

mvanderkamp commented:
Okay I've pushed a new attempt that just reuses the while-loop approach from `TreeInfo` since that was already doing the right thing. I thought about avoidng the while-loop by doing something with comparing the filename returned by `ls-files` against the requested path, but I figured the loop is actually fairly insignificant if run on just one line so may as well use it since it already does the right thing.

tpope commented:
You're building a whole data structure just to pluck one value out of it and throw the rest away. This is not a good approach, least of all from a code readability perspective. You should need little more than `if filename ==# path`.

mvanderkamp commented:
This is going to return the wrong information during merge conflicts, since it will always grab the first line, regardless of which stage was actually requested. Is there a way to only request one stage? I can't find it, and if not, we'll have to loop over the lines. I might take a stab at adding some caching back in at the same time.

mvanderkamp commented:
Okay, how does that look? Is the caching fine? In the original dicussion you suggested a more granular caching approach would be appropriate. I took that to mean using the repo and path as the cache key, giving a flatter structure overall.

It's a bit awkward having to fill in all four stages in the cache for tree objects- is there a better way to handle that?

tpope commented:
> Okay, how does that look? Is the caching fine? In the original dicussion you suggested a more granular caching approach would be appropriate. I took that to mean using the repo and path as the cache key, giving a flatter structure overall.

At least throw a `//` in the key. Don't create a bogus path like `/repo/.git/dir/file` that _looks_ real. There is also potential for collisions here.

How is performance without caching? Caching was basically mandatory back when we used `system()`, but now that we're on jobs I would be open to ditching it.

> It's a bit awkward having to fill in all four stages in the cache for tree objects- is there a better way to handle that?

Supporting `0` only is probably good enough. This whole fake tree object thing is a crude workaround for a broken abstraction anyways.

mvanderkamp commented:
Fetching an uncached entry:
```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    1   0.007091   0.000342  <SNR>130_IndexInfo()
```
Cached:
```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    1   0.001120   0.000068  <SNR>130_IndexInfo()
```
So roughly 6x to 7x faster. I don't have the experience to know what's generally acceptable for plugins. Personally I think 7 extra milliseconds isn't a big deal, so I'd honestly be fine with ditching the caching after all given the increased risk of bugs.

Reminder for myself as much as anyone: in its current form this function is only hit during merge conflicts, and then only when doing a `Gdiffsplit` on the files that are conflicting.

tpope commented:
Remind me, why doesn't it apply to `:Gdiffsplit` outside of a merge conflict?

mvanderkamp commented:
Sorry I deleted that part of the message. I wrote it down and then realized it sounded weird, tested it, found that actually it was just a bug in the caching, then somehow forgot to delete that part of the post.

tpope commented:
7ms is nigh imperceptible on its own, but that's going to be multiplied by number of checks, which in turn will be multiplied by levels of nesting. This can get out of hand quick. Even _with_ caching this could amount to a pretty big setback for smaller repos.

The Projectionist stuff isn't super high value and can even be confusing in some cases. I'm starting to get tempted to change it to opt-in rather than opt-out.

mvanderkamp commented:
Sorry for taking a while to get back to this! That makes sense that it wouldn't just be 7ms, I forgot that this is going to be hit repeatedly. That makes it sound like caching will be crucial to minimize the cost.

Here's what I've changed:
* I used the `//` separator as you suggested for the cache key, and also added a `cache://` prefix to really make it clear this isn't a normal path.
* Dropped support for stages 1-3 for tree objects.
* Fixed the bug that prevented this from working correctly on index files outside of merge conflicts.

Turning this functionality off by default seems reasonable to me.
